Batch Maps Feature Branch

gislawill opened the pull request:
### Description

This PR adds batch map export functionality (#1452), allowing users to generate multiple maps across a date range and download them as a merged PDF or ZIP of PNGs. The feature uses server-side rendering (playwright) to build and package PRISM rendered maps.

**Architecture**
The frontend's Print Dialog constructs URLs for each date in the selected range and sends them to a new /export API endpoint. The backend uses Playwright to render each URL via a dedicated /export frontend route (which reads all config from URL params), waits for window.PRISM_READY, then captures and packages the results.

```
Print Dialog (PRISM Frontend - image.tsx)
       │
       ▼ downloadBatch() constructs URLs
POST /export (PRISM API Server)
       │
       ▼ Playwright renders each URL
/export route (PRISM Frontend - ExportView)
       │
       ▼ Sets window.PRISM_READY when loaded
Returns merged PDF or ZIP
```

**Aspect Ratio Update**
This branch now includes the aspect ratio configuration feature (merged from #1648), which replaces the previous "Map Width" percentage slider with a more intuitive aspect ratio selector. Users can now choose from preset ratios, A4 print formats, a custom ratio input (w:h), or "Auto" mode which allows the map to resize based on the browser window. This approach provides both flexibility and consistency for **both** single and batch map exports.

## How to test the feature:
1. Start the app with a country that has date-supported layers: `REACT_APP_COUNTRY=mozambique yarn start`
4. Select a layer with date support (e.g., rainfall forecast)
5. Open Print Dialog (camera icon in navbar)
6. Configure appearance (title, legend position, etc.)
7. Enable batch maps: Toggle "Create a sequence of maps"
8. Select date range using the date pickers
9. Verify map count displays correctly
10. Download:
    * Click Download → "Download maps as PDF" → verify merged PDF with correct page count
    * Click Download → "Download maps as PNGs" → verify ZIP with correctly named PNG files

## Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please still tick them so we know you've gone through the checklist. -->

Test your changes with

- [ ] `REACT_APP_COUNTRY=rbd yarn start`
- [ ] `REACT_APP_COUNTRY=cambodia yarn start`
- [x] `REACT_APP_COUNTRY=mozambique yarn start`
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

## Screenshot/video of feature:
<img width="1506" height="1240" alt="Screenshot 2025-12-20 at 2 53 50 PM" src="https://github.com/user-attachments/assets/6601130e-376f-493e-9f0c-acef9f77b64d" />

gislawill commented:
@ericboucher could we get your review on the architecture and server-code for this feature? We'll need to deploy the server updates for the preview build to work and actually demonstrate the functionality. But before I do that, I'd love to get your thoughts.

I'll note that one item that's still in flux is how we manage the aspect ratio for the generated map. That is all being handled in #1648 based on this conversation in Slack https://ovio-public.slack.com/archives/C0958FADX2M/p1765328289971239

ericboucher commented on api/prism_app/export_maps.py:
```diff
@@ -0,0 +1,405 @@
+
+logger = logging.getLogger(__name__)
+
+# Timeouts
```
Should/could the playwright settings be shared with the report exports?

gislawill replied:
I don't have a great sense of how the report exports work. Are they triggered by something in the browser or only in CI? I think the BROWSER_LAUNCH_ARGS should likely be shared. Not sure about the rest of the set up

gislawill commented on api/prism_app/export_maps.py:
```diff
@@ -0,0 +1,405 @@
+    hostname_lower = hostname.lower()
+    is_allowed = False
+
+    for domain in EXPORT_ALLOWED_DOMAINS:
```
Yea, good call. I'll see if I can find library that does this

wadhwamatic commented:
> > UI tweaks :
> > How about using a dropdown for the aspect ratio? I think it's taking too much space atm. Same for the sequence map, how about moving the explanatory text to an info tooltip or similar?
> 
> I can make those changes… in my view the tradeoff we’d make for less space is that any aspect ratio change will take two clicks instead of one, and we’d lose the ability to visually show the ratios themselves and group options — @wadhwamatic do you have a preference on a dropdown vs. the toggle groups?

I do see the value in having a visual representation of the aspect ratios. We could tighten up the layout to reduce its footprint. We could have auto, a4 portrait, a4 landscape and custom as a single row of buttons. The text 'Map resizes based on browser window' could be a tooltip on hover over auto.  I think that alone would make a big difference. The sequence text could also be a tooltip over the batch toggle.

gislawill commented:
@wadhwamatic, thanks for the thorough QA! I'll add some thoughts across the different areas of feedback

1. Data not loading before printing
We added tracking to prevent this in [895b7f5](https://github.com/WFP-VAM/prism-app/pull/1633/commits/895b7f58e9065f90f836e5573f535690d75b4a81). Frustratingly, this is consistently preventing the issue in Moz but not our other deployments. I'll dig into why that is and update.
    - [ ] Fixed (will check when done)

2. Varying types of layers (not just WMS) with date support
I support limiting v1 to just WMS layers. FWIW I believe Cadre Harmonise should be trivial to support (may already be supported). I'm not exactly sure about the issue with the kobo set-up - the username and password in aws (created Oct 2023) isn't working unfortunately. Would you be able to send me some test credentials in Slack? I'm guessing we need to pass the credentials back to our browser rendered version for batch maps - a good follow-up feature.

3. Date selection 
Conor is getting to work on those updates. All of those updates/fixes seem very do-able.
    - [ ] Updated (will check when done)

4. Date clarity
When you say:
    > This feature highlights the need to better display the selected date in the map view. What if we added date as a variable in the title?

    I'm assuming you're referring to the generated map export (but correct me if not). Yes, I agree we could add it to the title. For UX, I see a couple of options:
     * Automatically add date to whatever title the user provides `<h1>{title} - {date.toLocaleDateString()}`
     * Give user a date substitution variable, like...
        <img width="952" height="355" alt="Screenshot 2025-12-23 at 1 32 10 PM" src="https://github.com/user-attachments/assets/6c2e5daa-af36-4f5f-b503-5a58323a7dbb" />

    For date formatting, we could use also use either YYYY-MM-DD or ([toLocaleDateString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString)). We could also support something more complex like [this templating language](https://day.js.org/docs/en/display/format). 

    Estimated level of effort for #1611: a couple hours. It seems pretty straightforward. Just let us know if you want us to pick this up.

wadhwamatic commented:
> * I support limiting v1 to just WMS layers. FWIW I believe Cadre Harmonise should be trivial to support (may already be supported). I'm not exactly sure about the issue with the kobo set-up - the username and password in aws (created Oct 2023) isn't working unfortunately. Would you be able to send me some test credentials in Slack? I'm guessing we need to pass the credentials back to our browser rendered version for batch maps - a good follow-up feature.

@gislawill - that's totally fine. let's limit to WMS layer for now as this is the real use case we have. Let's just disable the option for non-WMS layers. 

> I'm assuming you're referring to the generated map export (but correct me if not). Yes, I agree we could add it to the title. For UX, I see a couple of options:

Yes, that's what I meant, and I think that giving users a date substitution variable is the best approach. Let's use the existing date to locale functionality we have in date-utils.ts

> Estimated level of effort for https://github.com/WFP-VAM/prism-app/issues/1611: a couple hours. It seems pretty straightforward. Just let us know if you want us to pick this up.


Great, let's please address https://github.com/WFP-VAM/prism-app/issues/1611 while we're working on this feature.

ericboucher commented on api/prism_app/main.py:
```diff
@@ -544,3 +546,54 @@ def get_google_floods_inundations_api(
+        dates = extract_dates_from_urls(export_request.urls)
+        file_bytes, content_type = await export_maps(
+            urls=export_request.urls,
+            viewport_width=export_request.viewportWidth,
```
Maybe worth having defaults for viewport_width / viewport_height (standard full screen size?) so that we can easily request a map export with just a list of urls?

How do we control zoom / movements done in the frontend ?

gislawill commented:
> There is a grey line around the maps when batch downloading. This line is not present when using the standard download feature.

Good catch @ericboucher! I'm finding I actually prefer the small gray border on the current batch maps over the borderless single map. I'm going to make sure that the single map has the same border for consistency. Curious for @wadhwamatic's thoughts on whether there's a better design path here?

| Current batch map border | Current single map border (lack of) |
| - | - |
| <img width="1109" height="480" alt="Screenshot 2026-01-12 at 10 05 44 AM" src="https://github.com/user-attachments/assets/58d5f404-8408-40f7-b673-72d185210365" /> | <img width="1342" height="728" alt="Screenshot 2026-01-12 at 10 05 53 AM" src="https://github.com/user-attachments/assets/892a4fae-4e9c-4fc5-99e3-09affa90578c" /> |

gislawill commented on frontend/src/components/NavBar/PrintImage/mapDimensionsUtils.ts:
```diff
@@ -0,0 +1,118 @@
+/**
+ * All available aspect ratios with their decimal values
+ */
+const ASPECT_RATIOS: {
```
Good call, I've just created a new constants file that has a single source of truth for values + types

ericboucher commented:
@gislawill I am specifically talking about the border of the document, not the map itself

gislawill commented:
Removed document borders per thread in Slack (cc: @wadhwamatic & @ericboucher)
